lib/operations/add_user.js

```diff
@@ -0,0 +1,193 @@
+'use strict';
+
+const OperationBase = require('./operation').OperationBase;
+const CommandOperation = require('./command');
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const CONSTANTS = require('../constants');
+const count = require('./collection_ops').count;
+const crypto = require('crypto');
+const handleCallback = require('../utils').handleCallback;
+const MongoError = require('mongodb-core').MongoError;
+const ReadPreference = require('mongodb-core').ReadPreference;
+const toError = require('../utils').toError;
+const updateOne = require('./collection_ops').updateOne;
+
+let db;
+function loadDb() {
+  if (!db) {
+    db = require('../db');
+  }
+  return db;
+}
+
+class AddUserOperation extends OperationBase {
+  constructor(db, username, password, options) {
+    super(options);
+
+    this.db = db;
+    this.username = username;
+    this.password = password;
+  }
+
+  execute(callback) {
+    const db = this.db;
+    const username = this.username;
+    const password = this.password;
+    const options = this.options;
+
+    let Db = loadDb();
+
+    // Did the user destroy the topology
+    if (db.serverConfig && db.serverConfig.isDestroyed()) {
+      return callback(new MongoError('topology was destroyed'));
+    }
+    // Attempt to execute auth command
+    executeAuthCreateUserCommand(db, username, password, options, (err, r) => {
+      // We need to perform the backward compatible insert operation
+      if (err && err.code === -5000) {
+        const finalOptions = applyWriteConcern(Object.assign({}, options), { db }, options);
+
+        // Use node md5 generator
+        const md5 = crypto.createHash('md5');
+        // Generate keys used for authentication
+        md5.update(username + ':mongo:' + password);
+        const userPassword = md5.digest('hex');
+
+        // If we have another db set
+        const dbToUse = options.dbName ? new Db(options.dbName, db.s.topology, db.s.options) : db;
+
+        // Fetch a user collection
+        const collection = dbToUse.collection(CONSTANTS.SYSTEM_USER_COLLECTION);
+
+        // Check if we are inserting the first user
+        count(collection, {}, finalOptions, (err, count) => {
+          // We got an error (f.ex not authorized)
+          if (err != null) return handleCallback(callback, err, null);
+          // Check if the user exists and update i
+          const findOptions = Object.assign({ projection: { dbName: 1 } }, finalOptions);
+          collection.find({ user: username }, findOptions).toArray(err => {
```

Comment:
Just realized: does this require passing a session in?

Reply:
I'm not sure, does it?